telcell/data/utils.py

```diff
@@ -0,0 +1,62 @@
+import datetime
+from typing import Iterable, List, Tuple
+
+from telcell.data.models import Track
+
+
+def split_track_by_interval(
+        track: Track,
+        start: datetime.datetime,
+        end: datetime.datetime
+) -> Tuple[Track, Track]:
+    """
+    Splits the specified `track` in two separate `Track`s: one contains only
+    the `Measurement`s that fall within the specified [start, end) interval,
+    the other that contains all other measurements (i.e. those that occurred
+    before `start` or after `end`).
+
+    :param track: The track to be split in two
+    :param start: The start of the interval to split `track` on
+    :param end: The end of the interval to split `track` on (exclusive)
+    :return: Tracks of measurements within and outside the specified interval
+    """
+    selected_measurements = []
+    remaining_measurements = []
+
+    for measurement in track:
+        if start <= measurement.timestamp < end:
+            selected_measurements.append(measurement)
+        else:
+            remaining_measurements.append(measurement)
+
+    selected = Track(track.owner, track.name, selected_measurements)
+    remaining = Track(track.owner, track.name, remaining_measurements)
+    return selected, remaining
+
+
+def extract_intervals(
+        timestamps: Iterable[datetime.datetime],
+        start: datetime.datetime,
+        duration: datetime.timedelta,
+) -> List[Tuple[datetime.datetime, datetime.datetime]]:
+    """
+    Calculates, for a set of intervals and timestamps, which intervals have at
+    least one timestamp.
+
+    The first interval is defined by `start` and `duration`. The next interval
+    starts adjacent to (before or after) the first interval, and so on. An
+    interval is returned iff there is at least one timestamp in `timestamps`
+    that is within the interval.
+
+    :param timestamps: timestamps which determine which intervals are returned
+    :param start: the start of the first interval
+    :param duration: the duration of the intervals
+    :return: a sorted list of intervals for which there is at least 1 timestamp
+    """
+    intervals = set()
+    for ts in timestamps:
+        sequence_no = (ts - start) // duration
+        interval_start = start + sequence_no * duration
+        intervals.add((interval_start, interval_start + duration))
```

Comment:
Deze implementatie is misschien voor nu even goed genoeg, maar dit is de klassieke manier om ruzie te krijgen met DST-achtige constructies (als we volhouden dat we een dag zien als 05:00 tot 05:00, dan gaat dit ergens mis, toch?).

Reply:
Goed punt maar wat zou de oplossing zijn?

Reply:
Yeah die oplossing is wel lastig, enige wat ik me zou gauw kan verzinnen is als het gevraagde interval precies uitkomt op *n ≥ 1* dagen, dat we dat als signaal zouden gebruiken. Maar dat klinkt bij voorbaat al behoorlijk ingewikkeld...

Reply:
Wat mij betreft wel ja, kunnen we later wel een keer een ingewikkeld PR van maken :P